### PR TITLE
error messages: use de Bruijn indices for identifier disambiguation

### PR DESCRIPTION
**Pull request by Octachron**
Following the discussion in #11286, this PR proposes to have a simple and clear specification for disambiguated identifiers (t/2, t/3, ...) by using De Bruijn indice to disambiguate identifiers that share the same name. 

Currently, path disambiguation in error message has a complex and hard to document behavior.
For instance, if we define
```ocaml
type t = A;;
type t = B;;
type t = C;;
```
then if the type of `A` appears in an error message it might displayed as `t/2` or `t/3` depending on previous part of the error message.
For instance, both
```ocaml
let x, y = A, B in y = x;;
```
and
```ocaml
let x, y = A, B in x = y;;
```
prints the same error message
```
Error: This expression has type t/2 but an expression was expected of type
         t/3
```
while pointing to the right-hand side of the equality because they attribute the same disambiguated name to different identifiers.

With this PR, an identifier that has been shadowed `n` (n ≥ 1) time in the current scope will be displayed as `name/(n+1)`.
Consequently, the displayed name of identifiers is stable as long as one does not introduce a new identifier with the same name.

In the previous example, this means that the type of A is consistently displayed as `t/3`
```ocaml
A;;
```
```
- : t/3 = A
```

With this specification, it is no longer necessary to keep track of the naming decision for identifiers in the `Printtyp` module.
This reduces the complexity of the `Printtyp` module by removing one layer of global state. It would be also easier reuse this naming scheme for the illegal shadowing error messages which currently use the inner identifier stamp to avoid dealing with the complexity in `Printtyp`.

One potential drawback of this approach is that bad printing environment might fail to provide enough information to disambiguate some identifiers. This can currently happens in the functor error message. Indeed, the printing environment is not updated when we are printing the argument-by-argument suberror message. However, in the context of a functor error message, we often end up with a divergent environment between the left-hand side and right-hand side of the functor comparison. In this situation, it might be *better* to not over-eagerly disambiguate identifiers.
Morever, bad printing environment are already problematic for the short-paths. It thus seem reasonable to commit to having good printing environment.

### CODE REVIEW

**Comment by gasche** [comment]
I'm impressed that you dared implement this change, thanks! This is impressive.

For the ease of reviewing, could you maybe split the first commit between the implementation change and the testsuite update? It's a bit hard to find the code to review right now, and yes github provides a "file filter" feature but I'd rather not depend on it.

**Comment by Octachron** [comment]
I have split the first commit in two, you should now have a nice link to the implementation changes.

**Comment by xavierleroy** [comment]
Once again, de Bruijn indices save the day :-)

File: typing/env.ml [hunk 1]
```diff
@@ -3168,6 +3181,23 @@ let find_label_by_name lid env =
...
 
+(* Stable name lookup for printing *)
+
+let find_index_tbl ident tbl  =
+  let lbs = IdTbl.find_all_idents (Ident.name ident) tbl in
+  let find_ident (n,p) = match p with
+    | Some i -> if Ident.same ident i then Some n else None
+    | _ -> None
+  in
+  Seq.find_map find_ident @@ Seq.mapi (fun i x -> i,x) @@ List.to_seq lbs
```
Comment (gasche) [inline typing/env.ml]: Why not have `find_all_idents` return a `Seq.t` directly? In particular, this would allow to have constant-time work in the common case where the index is 0, while currently you traverse the whole table each time, potentially many times, right? (I'm surprised this does not show up as a time regression, but I guess our testsuite environments are very shallow and we never test our timings anyway.)

File: typing/env.ml [hunk 2]
```diff
@@ -3168,6 +3181,23 @@ let find_label_by_name lid env =
   let loc = Location.(in_file !input_name) in
   lookup_label ~errors:false ~use:false ~loc Projection lid env
 
+(* Stable name lookup for printing *)
+
+let find_index_tbl ident tbl  =
+  let lbs = IdTbl.find_all_idents (Ident.name ident) tbl in
+  let find_ident (n,p) = match p with
+    | Some i -> if Ident.same ident i then Some n else None
```
Comment (gasche) [inline typing/env.ml]: `i` suggests an integer variable, as you are using below, I would use another name

File: typing/env.mli [hunk 3]
```diff
@@ -268,6 +268,18 @@ val find_constructor_by_name:
...
+    of an identifier in a given environment. In other words, it returns how many
+    times an identifier has been shadowed by a more recent identifiers with the
+    same name in a given environment.
+*)
+val find_value_index:   Ident.t -> t -> int option
+val find_type_index:    Ident.t -> t -> int option
+val find_module_index:  Ident.t -> t -> int option
+val find_modtype_index: Ident.t -> t -> int option
+val find_class_index:   Ident.t -> t -> int option
+val find_cltype_index:  Ident.t -> t -> int option
```
Comment (gasche) [inline typing/env.mli:283]: It may be worth documenting that `None` can be returned when the identifier does not occur at all in the environment. In fact it is rather strange that the API would force us to consider this case, I would have expected it to be a precondition and get an `invalid_arg` exception in this case. (I see below that this is related to functor arguments diffing...)

File: typing/printtyp.ml [hunk 4]
```diff
@@ -55,7 +54,6 @@ let printing_env = ref Env.empty
    cmi present on the file system *)
 let in_printing_env f = Env.without_cmis f !printing_env
 
-let human_unique n id = Printf.sprintf "%s/%d" (Ident.name id) n
```
Comment (gasche) [inline typing/printtyp.ml:58]: I liked this abstraction. Now you have `"/"` twice in the code and we will have to change them both if we decide to go for `foo(n)` later. Why remove it?

File: typing/printtyp.ml [hunk 5]
```diff
@@ -133,15 +131,26 @@ module Conflicts = struct
...
+
+  let collect_explanation namespace id ~name =
     let root_name = Ident.name id in
-    if not (M.mem name !explanations) then
-      match Namespace.location namespace id with
-      | None -> ()
-      | Some location ->
-          let explanation = { kind = namespace; location; name; root_name } in
-          explanations := M.add name explanation !explanations
+    if root_name != name && not (M.mem name !explanations) then
```
Comment (gasche) [inline typing/printtyp.ml]: `!=` looks like a code smell. Why are we using it here?

File: typing/printtyp.ml [hunk 6]
```diff
@@ -133,15 +131,26 @@ module Conflicts = struct
...
+      begin
+        add namespace name id;
+        if not (M.mem root_name !explanations) then
+          match Namespace.lookup namespace root_name with
+          | Pident root_id -> add namespace root_name root_id
+          | exception Not_found | _ -> ()
+      end
```
Comment (gasche) [inline typing/printtyp.ml:160]: It is difficult to tell what this code is doing from reading it. It looks like its behavior changed in this patchset. Maybe a comment to explain what is going on with an example or two could help?

File: typing/printtyp.ml [hunk 7]
```diff
@@ -215,43 +224,9 @@ module S = String.Set
...
-let get namespace = map.(Namespace.id namespace)
-let set namespace x = map.(Namespace.id namespace) <- x
-
-(* Names used in recursive definitions are not considered when determining
-   if a name is already attributed in the current environment.
-   This is a complementary version of hidden_rec_items used by short-path. *)
-let protected = ref S.empty
+(* Names bound in recursive definitions should be considered as bound
+   in the environment when printing identifiers *)
+let bound_in_recursion = ref M.empty
```
Comment (gasche) [inline typing/printtyp.ml:257]: I forgot about this recursion business, and the definition of `trees_of_recursive_sigitems` lacks the comments that would help me remember what is going on.

From a high level perspective, I think it comes from the fact that we reconstruct typing environments as we traverse definitions for printing, but we reconstruct them incorrectly (not taking recursion into account?) and we patch things up after the fact with this trick.

With the previous codebase, sometimes when an identifier was "hidden" we would pretend that it did not already exist in the environment. (I suppose that this was used to ensure that we use `foo` instead of `foo/n` for this identifier, but I'm not sure.) (I also wonder if we can have these recursive-sigitems "hidden" additions interleaved with other actions acting on the environment and adding new possibly-conflicting identifiers. Can I enter a recursive group, then `open` a module that shadows this name, and enter another recursive group on a (differnt) identifier of the same name? Or are those hidden identifiers always the most recent additions, morally, to the environment?)

With the current codebase, when an identifier is "hidden" it results in a shift in the De Bruijn index computation, which is simpler and easier to understand than what we had before (but would still be worth having an explanation with an example), and seems to suggest that we assume that those "hidden" identifiers are always the most recent declaration in context, not interleaved with other same-name declarations.

File: typing/printtyp.ml [hunk 8]
```diff
@@ -263,99 +238,61 @@ let with_arg id f =
...
+       Printing the source-level name seems like a reasonable compromise
+       in this situation however.*)
+  | None, Some _ | Some 0, None -> Ident.name id
+  | Some n, Some p ->
+      if Ident.same p id then
+        Ident.name id
+      else
+        String.concat "/" [Ident.name id; string_of_int (2 + n)]
+  | Some n, None ->
+      String.concat "/" [Ident.name id; string_of_int (1 + n)]
```
Comment (gasche) [inline typing/printtyp.ml]: Why `(1+n)`, `(2+n)`? I thought that `n=0` meant "this is the last declaration in the environment", and we would show `foo` instead of `foo/0`, But then `n=1` should mean "this is the last declaration *shadowed* in the environment" (there is only `foo` after it) and I would expect `foo/1` in this case, right? Or did you consider that `foo/2` for this was more natural as the "second" definition? I think this choice could be explained somewhere. (We could have a function to encapsulate "how we show this ambiguous identifiers to a human", oh wait...)

File: typing/printtyp.ml [hunk 9]
```diff
@@ -263,99 +238,61 @@ let with_arg id f =
...
+       within functor error messages where the left and right hand side
+       have a different views of the environment at the source level.
+       Printing the source-level name seems like a reasonable compromise
+       in this situation however.*)
+  | None, Some _ | Some 0, None -> Ident.name id
+  | Some n, Some p ->
+      if Ident.same p id then
+        Ident.name id
+      else
+        String.concat "/" [Ident.name id; string_of_int (2 + n)]
```
Comment (gasche) [inline typing/printtyp.ml]: I would find the handling of `rec_bound` easier to understand if it was formulated as an information used to transform/fix the indexed name, rather than a parallel information used for printing in addition to the indexed name. (This comes from the view that rec_bound business should not be necessary if the environment correctly had those extra bindings in it, so we are fixing the environment rather than changing the printing.) This would in fact be similar to how `env_ident` was designed in the previous version.

Sketch:

```ocaml
let index = in_printing_env (find namespace id) in
let index =
  (* The environment we are passed does not countain recursive occurrences,
     which should have precedence if they exist. Fix the index if necessary. *)
  match M.find_opt (Ident.name id) !bound_in_recursion with
  | None -> index
  | Some rec_ident ->
    if Ident.same rec_ident id
    then
      (* we are in fact a local recursive occurrence *)
      Some 0
    else
      (* we shadow an additional local recursive occurrence *)
      Some (1 + n)
in
match index with
| None -> (* this case ... *) Ident.name id
| Some n -> human_name id n
```

File: typing/printtyp.ml [hunk 10]
```diff
@@ -413,22 +350,31 @@ let rewrite_double_underscore_paths env p =
...
+      Oide_apply (t1, t2)
+
+let tree_of_path ?disambiguation namespace p =
+  tree_of_path ?disambiguation namespace
+    (rewrite_double_underscore_paths !printing_env p)
 
-let tree_of_path namespace p =
-  tree_of_path namespace (rewrite_double_underscore_paths !printing_env p)
+let best_tree_of_type_path p p' =
+  if p == p' then tree_of_path Type p'
```
Comment (gasche) [inline typing/printtyp.ml]: I'm nervous again here. For me physical equality should be used as an optimization to shortcut a more expansive (but more meaningful) equality check. When you do this, you know the physical check is always an optimization that can b removed without changing the behavior -- if we remove the short path, we end up in the equality check that gives the same result. It is not clear that it is the case here.

File: typing/printtyp.ml [hunk 11]
```diff
@@ -413,22 +350,31 @@ let rewrite_double_underscore_paths env p =
...
       (* [t.A]: inline record of the constructor [A] from type [t] *)
       Oide_dot (Oide_ident (ident_name Type t), s)
   | Pdot(p, s) ->
-      Oide_dot (tree_of_path Module p, s)
+      Oide_dot (tree_of_path ~disambiguation Module p, s)
   | Papply(p1, p2) ->
-      Oide_apply (tree_of_path Module p1, tree_of_path Module p2)
+      let t1 = tree_of_path ~disambiguation Module p1 in
+      let t2 = tree_of_path ~disambiguation Module p2 in
+      Oide_apply (t1, t2)
```
Comment (gasche) [inline typing/printtyp.ml:414]: You can reduce the diff with

```ocaml
let rec tree_of_path ?(disambiguation=true) namespace p =
  let tree_of_path namespace p = tree_of_path ~disambiguation namespace p in
  let namespace = if disambiguation then namespace else Other in
  match p with
```

**Review by gasche — COMMENTED** [review]
This looks like a nice improvement overall: the spec is simpler and the implementation as well. But still there are a bit too many moving pieces in the printtyp codebase for me to get a complete understanding in one go. So this is just a first round of questions, I guess.

File: typing/env.ml (hunk 1 above)
Comment (Octachron) [reply typing/env.ml]: I am in favor of switching to a `Seq.t` but I am not certain if the optimization is necessary: this is a printing only function, and I would expect that types are not shadowed deeply and frequently enough for the difference to matter in printing.

File: typing/env.mli (hunk 3 above)
Comment (Octachron) [reply typing/env.mli:283]: I can add the documentation.

I mostly kept this current API as a canary that there is still something odd going on in some cases which will require some design choice at a later stage.

But yes ultimately, the fact that the identifier being printed belongs to the printing environment should be a precondition.

File: typing/env.ml (hunk 1 above)
Comment (gasche) [reply typing/env.ml]: I was probably too pessimistic above, but I wonder if we should worry about, say, Merlin printing a larger-than-average type in the middle of a program with several `open` for modules that, by convention, very often define a type named `t`. In any case `Seq.t` is not harder to write and eliminates this worry completely.

File: typing/printtyp.ml (hunk 7 above)
Comment (Octachron) [reply typing/printtyp.ml:257]: My high-level explanation is that when printing recursive definitions there are two printing environment that are natural to consider: the one just before the recursive definitions and the one with the recursive components bound in the scope. The short-paths functions work more naturally on the `just-before` environment because the identifiers are not yet valid shorthand for longer names, 
```ocaml
module Avoid__me = struct
  type t = A
end
type u = Avoid__me.t
```
whereas the `name conflict` functions need to take in account the environment extended with the current recursive identifiers which are already in the environment:
```ocaml
type t = A
type u = [ `A of t ]
module M = struct
  type t = [ u | `B of t ]  
end
```

Similarly, only new definition can introduce new identifiers in the environment that can shadow recursive identifiers.
Concerning environment manipulation, indeed there is a bug in the printer in the case of recursive module where it is possible to shadow the recursive module being defined by a non-recursive one:
```ocaml
module rec M: sig
  type t
  type r = [`A of M.t]
  module M: sig
     type t
   end
  type s = [ `B | r ]
end = struct
  type t
  type r = [`A of M.t]
  module M = struct
     type t
   end
  type s = [ `B | r ]
end
```
Here the printer expand the definition of `s` into
```ocaml
    type s = [ `A of M.t | `B ]
```
which refers to the shadowed module `M` directly.

Overall, I think that the best fix is to switch to a second explicit environment rather than rely on the ad-hoc `bound_in_recursion` type.

File: typing/printtyp.ml (hunk 8 above)
Comment (Octachron) [reply typing/printtyp.ml]: I had indeed opted for `2` as the most recognizable second ordinal rather than `1`. And yes having a documented function is probably a good idea.

File: typing/printtyp.ml (hunk 10 above)
Comment (Octachron) [reply typing/printtyp.ml]: The simplest fix here is probably to ensure that the `short-path` function returns directly the information that it has returned an alternative path or not.

File: typing/printtyp.ml (hunk 6 above)
Comment (Octachron) [reply typing/printtyp.ml:160]: The apparent change of behavior is due to the fact that we no longer implicitly (re-)name the identifier in the scope with `env_ident`. Consequently, the `collect_explanation` function might no longer be called directly on the identifier in scope. This new part of the code ensure that whenever we register a conflicting identifier, we do add the identifier-in-scope to the set of conflicting identifiers.

File: typing/printtyp.ml (hunk 7 above)
Comment (gasche) [reply typing/printtyp.ml:257]: It's okay if the code is not completely correct on first change -- incremental improvements are fine. But it would be nice to have comments (as you just wrote) to explain what the code is trying to be doing.

File: typing/printtyp.ml [hunk 12]
```diff
@@ -225,7 +231,28 @@ let enabled = ref true
...
+     type t = A
+   end
+   type t = X
+   type u = t * t
+   module M = struct
+     type t = A of u
+     type r = Avoid__me.t
+   end
+  }]
+  It is is important that in the definition of [t] that the outer type [t] is
```
Comment (gasche) [inline typing/printtyp.ml:249]: of [M.t] ? ("that in the definition of [M.t] ~~that~~ the outer type ...").

Comment (gasche) [reply typing/printtyp.ml:249]: Naive question: why would we print "the outer type [t]" (is this `t = A` or `t = X`?) "in the definition of `t`"? The definition of A contains `u` but not `t`. Would this `u` be unfolded by the printer?

File: typing/printtyp.ml [hunk 13]
```diff
@@ -225,7 +231,28 @@ let enabled = ref true
...
+  printed as [t/2] reserving the name [t] to the type being defined in the
+  current recursive definition.
+     Contrarily, in the definition of [r], one should not shorten the
+  path [Avoid__me.t] to [r] until the end of the definition of [r].
+  The [bound_in_recursion] bridges the gap between those two slightly different
+  notions of printing environment.
```
Comment (gasche) [inline typing/printtyp.ml:255]: If I understand correctly, we get a problem if, working on the occurrence of `Avoid__me.t` in the definition of `M.r`, we would call short-path with an environment where `r` is already in scope and bound to `Avoid__me.t`. Naive idea: could we use an environment where `r` is bound but abstract? Would this allow to use a single notion of environment for both use-cases?

Generally the idea is that in `let rec x = t in u`, you would use for `t` an environment with `x` is bound but abstract, and you would use the full definition `x = t` only in `u`.

(This assumes there is a notion of "abstract" for each kind of objects that can be recursively defined and are shown by printtyp. I was going to make a joke on abstract module types saving the day again, but I don't know how to define a recursive module type anyway!)

File: typing/printtyp.ml [hunk 14]
```diff
@@ -242,6 +269,18 @@ let with_hidden ids f =
   let updated = List.fold_left update !bound_in_recursion ids in
   protect_refs [ R(bound_in_recursion, updated )] f
 
+let human_id id index =
+  if index = 0 then
+    Ident.name id
+  else
+  (* We print the most k-time overshadowed identifier as `name/(k+1)`.
```
Comment (gasche) [inline typing/printtyp.ml]: I'm not sure what "most k-time overshadowed" means.

I think you could have a shorter explanation anyway: humans are more comfortable with 1-indexing instead of 0-indexing so we do that, and we show `foo` instead of `foo/1`.

(I'm not sure I like this choice to be honest. It is natural to think of `foo/n` as "the (1-indexed) n-th shadowed definition", so `foo/1` is the first shadowed definition so it is different from `foo`.)

File: typing/printtyp.ml [hunk 15]
```diff
@@ -251,25 +290,32 @@ let indexed_name namespace id =
...
+          Some 0
+        else
+          (* the current recursive definition shadows one more time the
+            previously existing identifier with the same name *)
+          Option.map succ (in_printing_env (find namespace id))
+    | None ->
+        in_printing_env (find namespace id)
+  in
+  let index =
+    (* The  [None] case is potentially problematic, it might indicate that
```
Comment (gasche) [inline typing/printtyp.ml]: I would use: "If [index] is [None] at this point, it might indicate..."

File: typing/printtyp.ml [hunk 16]
```diff
@@ -711,6 +754,13 @@ let best_type_path p =
     (* Format.eprintf "%a = %a -> %a@." path p path p' path p''; *)
     (p'', s)
 
+(* When building a tree for a best type path, we should not disambiguate
+   identifiers whenever the short-path algorithm detected a better path than
+   the original one.*)
+let tree_of_best_type_path p p' =
```
Comment (gasche) [inline typing/printtyp.ml:769]: nice renaming!

**Review by gasche — APPROVED** [review]
Thanks! I like the new version. I left some comments on the comments you wrote of course. I am not fond of your choice of human-printing strategy, see inline comment. This is still worth a preemptive approval.

**Comment by gasche** [comment]
@Octachron in https://github.com/ocaml/ocaml/issues/6634#issuecomment-1286642883 you show an example message:

```
Error: Signature mismatch:
       Modules do not match:
         sig type t = [ `T of t ] end
       is not included in
         sig type t = [ `T of t/2 ] end
       Type declarations do not match:
         type t = [ `T of t ]
       is not included in
         type t = [ `T of t/3 ]
       The type [ `T of t ] is not equal to the type [ `T of t/2 ]
       Type t = [ `T of t ] is not equal to type t/2 = int
       Types for tag `T are incompatible
```

why are we seeing ``type t = [ `T of t/3 ]`` in the middle of the thing? My guess is that `t/2` and `t/3` are actually the same object, but the error-message explanation logic is printing them in different environments. Is this avoidable? (This suggests that De Bruijn indices, while simpler than weirdly-cached-mutable-state decisions and more robust in theory, also add their own sort of fragility in that they expose the semi-rigorous environment-passing behavior of the internal implementation.)

File: typing/printtyp.ml (hunk 12 above)
Comment (Octachron) [reply typing/printtyp.ml:249]: Hm, I forgot to update the example to the really problematic case:
```ocaml
type u = [ `A of t * t ]
module M = struct
  type t = A of [ u | `B ]
  ...
end
```
where the type definition of `u` is inlined inside the definition of `t`.

File: typing/printtyp.ml (hunk 13 above)
Comment (Octachron) [reply typing/printtyp.ml:255]: That might work with the current implementation of `short-path`.

Comment (gasche) [reply typing/printtyp.ml:255]: Who cares about the other implementation? (If someone did, they would submit it upstream, right?)

**Comment by Octachron** [comment]
In this case, we are (temporarily?) exposing one of the short-path security belt that hides the type being declared by adding yet another type `t` to the printing environment: https://github.com/ocaml/ocaml/blob/trunk/typing/printtyp.ml#L2539 .

File: typing/printtyp.ml (hunk 13 above)
Comment (Octachron) [reply typing/printtyp.ml:255]: I am pretty sure that I am part of the set of people caring. And replacing the current algorithm would be a good time to remove the friction between `short-path` and the rest of `Printtyp`.

File: typing/printtyp.ml (hunk 14 above)
Comment (Octachron) [reply typing/printtyp.ml]:
```
The identifier with index [k] is the (k+1)-th most recent identifiers in
the printing environment. We print it as [name/(k+1)], except for [k=0]
which is printed as [name] rather than [name/1].
```
?

Comment (gasche) [reply typing/printtyp.ml]: 'most recent identifier' rather than 'most recent identifiers'

File: typing/printtyp.ml [hunk 17]
```diff
@@ -2522,6 +2521,3 @@ let tree_of_path = tree_of_path Other
 let tree_of_modtype = tree_of_modtype ~ellipsis:false
 let type_expansion mode ppf ty_exp =
   type_expansion ppf (trees_of_type_expansion mode ty_exp)
-let tree_of_type_declaration ident td rs =
-  with_hidden_items [{hide=true; ident}]
-    (fun () -> tree_of_type_declaration ident td rs)
```
Comment (gasche) [inline typing/printtyp.ml]: I'm not sure why this change is included here, and don't know if it is correct. Is it intentional?

Comment (Octachron) [reply typing/printtyp.ml]: No that was a leftover from my tests to check if it was really necessary to also hide the identifier in this case. This has been fixed.

**Review by gasche — APPROVED** [review]
I have reviewed this thoroughly in the part and I believe that the PR is now as good as one can reasonably expect it to be. Approved.

I believe that @Octachron and myself still disagree on the best numbering schemes for shadowed identifiers.
- I would prefer the first shadowed identifier for `foo` to be called `foo/1`, then `foo/2` etc., with `foo/<n>` interpreted as "the n-times-shadowed `foo`" (note that `foo`, the current identifier in scope, is not shadowed at all, so equivalent to `foo/0` which we never display to users).
- @Octachron chose to have first shadowed identifier for `foo` to be `foo/2`, intentionally leaving a gap with `foo/1` (or `foo/0`) never shown to users. `foo/<n>` is interpreted as "the n-th definition of `foo` of the environment" (for the natural-language notion of n-th starting at 1), so `foo` is equivalent to `foo/1`.

I find it slightly confusing to see `foo/2` but never `foo/1`, but I am happy to live with this disagreement and to let @Octachron have the final choice on this, he did all the work in the first place.
